### PR TITLE
Context

### PR DESCRIPTION
**Pull request by lukejacksonn**
Moves all state into a single context bound to a single reducer. The rational here is having all state in one place is going make the app more maintainable going forward. Given there is not much state and not many views, the chance of perf degradation is low and could be patched with some memoization if needs be.

The global state is available via hook in any function component:

```js
const [state, dispatch] = useStateValue();
``` 
The global state can be updated using dispatch:

```js
dispatch({ type, payload });
```

### CODE REVIEW

File: components/FileOverview.js [hunk 1]
```diff
@@ -1,87 +1,58 @@
...
-        packageName=${`${name}@${version}`}
-        filter=${searchTerm}
+  const file = cache['https://unpkg.com/' + request.path];
+  return (
+    !!file &&
+    html`
+      <${SearchInput}
+        placeholder="Search for dependencies.."
+        value=${searchTerm}
+        onChange=${setSearchTerm}
```
Comment (kiraarghy) [inline components/FileOverview.js]: I'm assuming we don't want this to be saved when switching between views?

File: components/PackageOverview.js [hunk 2]
```diff
@@ -1,54 +1,55 @@
...
-export const PackageOverview = ({ file, versions = [] }) => {
-  const [searchTerm, setSearchTerm] = react.useState('');
-  const {
-    pkg: { name, version, description },
-    meta,
-  } = file;
-
+export const PackageOverview = () => {
+  const [
+    { versions, request, directory, fileSearchTerm },
```
Comment (kiraarghy) [inline components/PackageOverview.js]: What is this array then object destructuring pattern, I've not seen it before?

File: components/FileOverview.js [hunk 3]
```diff
@@ -1,87 +1,58 @@
...
-        filter=${searchTerm}
-      />
-    </div>
-  `;
+      <div className=${styles}>
+        <div key="filesize">
+          <h2>File Size</h2>
+          <small>${formatBytes(file.size)}</small>
+        </div>
+        <${FileList}
```
Comment (alex-saunders) [inline components/FileOverview.js]: Could we render this conditionally based on `file.dependencies.length`? Currently, it 'Dependencies' title renders even if there are no files to display

File: components/Main.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+export default () => {
+  const [state, dispatch] = useStateValue();
+  const { request, code, packagesSearchTerm } = state;
+
+  // Update the request on user navigation
+  react.useEffect(() => {
+    const updateRequest = () => {
+      dispatch({ type: 'setRequest', payload: parseUrl() });
+    };
+    addEventListener('popstate', updateRequest);
```
Comment (alex-saunders) [inline components/Main.js]: As best practice, can we clean up this event listener as the return of this `useEffect`?

File: components/Main.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  // Fetch package meta data for all versions
+  react.useEffect(() => {
+    if (request.name)
+      fetch(`https://registry.npmjs.cf/${request.name}/`)
+        .then(res => res.json())
+        .then(json => dispatch({ type: 'setVersions', payload: json.versions }))
+        .catch(console.error);
+  }, [request.name]);
+
+  // Fetch package meta data for all versions
```
Comment (alex-saunders) [inline components/Main.js]: This comment looks wrong

File: components/Nav.js [hunk 6]
```diff
@@ -1,36 +1,37 @@
...
-          ? html`
-              <${FileOverview} file=${file} dispatch=${dispatch} />
-            `
-          : null)}
+      <${RadioGroup}
+        options=${modeOptions}
+        onClick=${val => dispatch({ type: 'setMode', payload: val })}
+      />
+      ${mode === 'package'
+        ? html`
```
Comment (kiraarghy) [inline components/Nav.js]: I would prefer this to be a switch but also I'm the only member of the switch lobby so 🤷‍♀

File: components/FileOverview.js (hunk 1 above)
Comment (lukejacksonn) [reply components/FileOverview.js]: We probably should for consistency, will add it to the globalState 👍

File: components/FileOverview.js (hunk 3 above)
Comment (lukejacksonn) [reply components/FileOverview.js]: When there are no files to show it says "0 Files" which perhaps isn't obvious enough by the sounds of it, but somewhat justifies keeping the label I reckon.

Comment (alex-saunders) [reply components/FileOverview.js]: Ah yeah sorry I missed the "0 FIles" string. I think this is fine for now.

**Pull request merged**